docker/indexer/stages/processing/processing.go

```diff
@@ -123,7 +144,54 @@ func (s *Stage) processGit(ctx context.Context, repoInfo *preparation.Result) er
 		}
 		return nil
 	}); err != nil {
-		return err
+		return fmt.Errorf("failed during file walk: %v", err)
 	}
-	return s.Storer.Store(ctx, repoInfo, shared.MD5, fileResults)
+
+	log.Info("begin processing tree")
+	treeResults, bucketResults := processBuckets(fileResults)
+	log.Info("begin storage")
+	return s.Storer.Store(ctx, repoInfo, shared.MD5, bucketResults, treeResults)
+}
+
+// Returns bucket hashes and the individual file hashes of each bucket
+func processBuckets(fileResults []*FileResult) ([]*BucketNode, [][]*FileResult) {
+	buckets := make([][]*FileResult, bucketCount)
+
+	for _, fr := range fileResults {
+		// Evenly divide into bucketCount buckets,
+		idx := binary.BigEndian.Uint16(fr.Hash[0:2]) % bucketCount
+		buckets[idx] = append(buckets[idx], fr)
+	}
+
+	results := make([]*BucketNode, bucketCount)
+
+	for bucketIdx := range buckets {
+		// Sort hashes to produce deterministic bucket hashes
+		sort.Slice(buckets[bucketIdx], func(i, j int) bool {
+			for k := 0; k < len(buckets[bucketIdx][i].Hash); k++ {
+				if buckets[bucketIdx][i].Hash[k] < buckets[bucketIdx][j].Hash[k] {
```

Comment:
Can we just use https://pkg.go.dev/bytes#Compare here instead for lexicographic comparison of these hashes?

Reply:
Yep, replaced it with bytes.Compare and added a test.